substrate/pkg/controller/substrate/cluster/instance.go

```diff
@@ -20,68 +20,109 @@ import (
+	if len(substrate.Status.PublicSubnetIDs) == 0 || substrate.Status.Cluster.LaunchTemplateVersion == nil {
 		return reconcile.Result{Requeue: true}, nil
 	}
-	if _, err := a.AutoScaling.CreateAutoScalingGroupWithContext(ctx, &autoscaling.CreateAutoScalingGroupInput{
-		AutoScalingGroupName: discovery.Name(substrate),
-		DesiredCapacity:      aws.Int64(1), MaxSize: aws.Int64(1), MinSize: aws.Int64(1),
-		LaunchTemplate:    &autoscaling.LaunchTemplateSpecification{LaunchTemplateName: discovery.Name(substrate), Version: aws.String("$Latest")},
-		VPCZoneIdentifier: aws.String(strings.Join(substrate.Status.PublicSubnetIDs, ",")),
-		Tags: []*autoscaling.Tag{{
-			Key:               aws.String(discovery.OwnerTagKey),
-			Value:             aws.String(substrate.Name),
-			PropagateAtLaunch: aws.Bool(true),
-		}, {
-			Key:               aws.String("Name"),
-			Value:             discovery.Name(substrate),
-			PropagateAtLaunch: aws.Bool(true),
-		}},
-	}); err != nil {
-		if strings.Contains(err.Error(), "Invalid IAM Instance Profile name") { // Instance Profile can take several seconds to propagate
+	instancesOutput, err := i.EC2.DescribeInstancesWithContext(ctx, &ec2.DescribeInstancesInput{})
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("describing instances, %w", err)
+	}
+	for _, reservation := range instancesOutput.Reservations {
+		for _, instance := range reservation.Instances {
+			if aws.StringValue(instance.State.Name) == ec2.InstanceStateNameRunning && aws.StringValue(instance.State.Name) == ec2.InstanceStateNamePending {
+				for _, tag := range instance.Tags {
+					if aws.StringValue(tag.Key) == "aws:ec2launchtemplate:version" && aws.StringValue(tag.Value) == aws.StringValue(substrate.Status.Cluster.LaunchTemplateVersion) {
+						logging.FromContext(ctx).Infof("Found instance %s", aws.StringValue(instance.InstanceId))
+						return reconcile.Result{}, nil
+					}
+				}
+			}
+		}
+	}
+	overrides := []*ec2.FleetLaunchTemplateOverridesRequest{}
+	for _, subnet := range substrate.Status.PublicSubnetIDs {
+		overrides = append(overrides, &ec2.FleetLaunchTemplateOverridesRequest{SubnetId: aws.String(subnet)})
+	}
+	createFleetOutput, err := i.EC2.CreateFleetWithContext(ctx, &ec2.CreateFleetInput{
+		Type: aws.String(ec2.FleetTypeInstant),
+		LaunchTemplateConfigs: []*ec2.FleetLaunchTemplateConfigRequest{{
+			Overrides: overrides,
+			LaunchTemplateSpecification: &ec2.FleetLaunchTemplateSpecificationRequest{
+				LaunchTemplateName: discovery.Name(substrate),
+				Version:            substrate.Status.Cluster.LaunchTemplateVersion,
+			}},
+		},
+		TargetCapacitySpecification: &ec2.TargetCapacitySpecificationRequest{
+			DefaultTargetCapacityType: aws.String(ec2.DefaultTargetCapacityTypeOnDemand),
+			TotalTargetCapacity:       aws.Int64(1),
+		},
+		TagSpecifications: discovery.Tags(substrate, ec2.ResourceTypeInstance, discovery.Name(substrate)),
```

Comment:
Does fleet automatically adds the launch template version tag `"aws:ec2launchtemplate:version"` to the instance?

Reply:
Yep!